site/source/docs/porting/connecting_cpp_and_javascript/embind.rst

```diff
@@ -344,31 +349,96 @@ The JavaScript code does not need to worry about lifetime management.
 Advanced class concepts
 =======================
 
+.. _embind-object-ownership:
+
+Object Ownership
+----------------
+
+JavaScript and C++ have very different memory models which can lead to it being
+unclear which language owns and is responsible for deleting an object when it
+moves between languages. To make object ownership more explicit, *embind*
+supports smart pointers and return value policies. Return value
+polices dictate what happens to a C++ object when it is returned to JavaScript.
+
+To use a return value policy, pass the desired policy into function or method
+bindings. For example:
+
+.. code:: cpp
+
+    EMSCRIPTEN_BINDINGS(module) {
+      function("createData", &createData, return_value_policy::take_ownership());
+    }
+
+Embind supports three return value policies that behave differently depending
+on the return type of the function. The policies work as follows:
+
+* *default (no argument)* - For return by value and reference a new object will be allocated using the
+  object's copy constructor. JS then owns the object and is responsible for deleting it. Returning a
+  pointer is not allowed by default (use an explicit policy below).
```

Comment:
Hmm, even with the new text I'm confused here. This is probably my own fault :smile: But say we have a function that returns by value,
```cpp
Object getObject() { .. }
```

What is, effectively, the bindings code for that? The text here says the copy constructor is called, so I am imagining
```cpp
void* bind_getObject() {
  Object copy = getObject(); // copy constructor occurs here during the `=` operation
  ...
}
```
And somehow in the `...` we transfer ownership to JS. But how does `copy` not end up collected by C++ when that scope ends?

To put my question another way, how can you call the copy constructor without opting into automatic memory management?

Reply:
The copy constructor is explicitly called with a `new`. e.g. the code for the above would be:
```c++
void* bind_getObject() {
  return new Object(getObject());
}
```
Or in terms of the embind, that happens in the `toWireType` function [here](https://github.com/emscripten-core/emscripten/pull/21692/files#diff-9e4cc1973ede37b961d9f900e3abd6befa0e62eba7d51a8229d49a3a3ff31074R364).

Reply:
Ah, thanks, so it's a raw `new` that uses the copy constructor. Makes sense to me now.